### PR TITLE
Add support for 'Recent Reciters' feature

### PR DESCRIPTION
**Pull request by zukhan**
recent_reciters

### CODE REVIEW

File: Sources/QuranAudioKit/Preferences/AudioPreferences.swift [hunk 1]
```diff
@@ -7,19 +7,50 @@
...
     @Preference(lastSelectedReciterId)
     public var lastSelectedReciterId: Int
+
+    @TransformedPreference(recentReciterIds, transformer: recentReciterIdsTransfomer)
+    public var recentReciterIds: OrderedSet<Int>
+
+    public func recentReciters(_ allReciters: [Reciter]) -> [Reciter] {
+        var recentReciters: [Reciter] = []
+        for recentReciterId in self.recentReciterIds {
+            recentReciters.append(allReciters.filter { $0.id == recentReciterId }[0])
```
[the remark this answers is not in the document]
Comment (zukhan) [reply Sources/QuranAudioKit/Preferences/AudioPreferences.swift:40]: Added a test case for the case when there are no recent reciters. It doesn't crash because it never goes inside the `for recentReciterId in self.recentReciterIds` loop (filter never gets called). 

This method is called by the App code (there is a corresponding PR in the private repo that's calling this method).

File: Sources/QuranAudioKit/Preferences/AudioPreferences.swift [hunk 2]
```diff
@@ -7,19 +7,50 @@
...
+        recentReciters.reverse()
+        return recentReciters
+    }
+
+    public func updateRecentRecitersList(_ reciter: Reciter) {
+        // Remove from the set if it exists in the set so it can go to the front (recently selected)
+        recentReciterIds.remove(reciter.id)
+
+        // Remove the least recently selected reciter if the list contains max reciters
+        if recentReciterIds.count >= Self.maxNumOfRecentReciters {
```
[the remark this answers is not in the document]
Comment (zukhan) [reply Sources/QuranAudioKit/Preferences/AudioPreferences.swift:51]: `Self.maxNumOfRecentReciters` is just a constant that's initialized at the top of the class to 3. According to the [Swift docs](https://docs.swift.org/swift-book/LanguageGuide/TheBasics.html#//apple_ref/doc/uid/TP40014097-CH5-XID_411), Int is preferred:

> Use UInt only when you specifically need an unsigned integer type with the same size as the platform’s native word size. If this isn’t the case, Int is preferred, even when the values to be stored are known to be nonnegative. A consistent use of Int for integer values aids code interoperability, avoids the need to convert between different number types, and matches integer type inference, as described in [Type Safety and Type Inference](https://docs.swift.org/swift-book/LanguageGuide/TheBasics.html#ID322).

File: Sources/QuranAudioKit/Preferences/AudioPreferences.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (AdibContractorCC) [reply Sources/QuranAudioKit/Preferences/AudioPreferences.swift:40]: In that case, write a test where self.recentReciterIds contains an id that is not present in allReciters - i.e. where filter will execute and return an empty array.

File: Sources/QuranAudioKit/Preferences/AudioPreferences.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (AdibContractorCC) [reply Sources/QuranAudioKit/Preferences/AudioPreferences.swift:51]: since this app is completely in swift, we don't need to worry about interoperability. Regardless, we can leave it Int, but it's safer to check that count is greater than 0 before trying to remove at that index.

[the remark this answers is not in the document]
Comment (AdibContractorCC) [reply Sources/QuranAudioKit/Preferences/AudioPreferences.swift:51]: alternatively, make an extension of the Array class to add methods that safely result in no-op for bad input and don't throw errors that can crash the app in prod

File: Sources/QuranAudioKit/Preferences/AudioPreferences.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (zukhan) [reply Sources/QuranAudioKit/Preferences/AudioPreferences.swift]: Added the test for the edge case. It shouldn't be possible for a reciter ID that's not in allReciters to end up on the recentReciterIds but in case it happens, I've changed it to what you suggested so it doesn't crash.

**Review by AdibContractorCC — APPROVED** [review]

File: Sources/QuranAudioKit/Preferences/AudioPreferences.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (zukhan) [reply Sources/QuranAudioKit/Preferences/AudioPreferences.swift]: Added a `!recentReciterIds.isEmpty` condition

File: Sources/QuranAudioKit/Preferences/AudioPreferences.swift [hunk 3]
```diff
@@ -7,19 +7,52 @@
...
     @TransformedPreference(audioEndKey, transformer: .rawRepresentable(defaultValue: .juz))
     public var audioEnd: AudioEnd
 
     @Preference(lastSelectedReciterId)
     public var lastSelectedReciterId: Int
+
+    @TransformedPreference(recentReciterIds, transformer: recentReciterIdsTransfomer)
+    public var recentReciterIds: OrderedSet<Int>
+
+    public func recentReciters(_ allReciters: [Reciter]) -> [Reciter] {
```
Comment (mohamede1945) [inline Sources/QuranAudioKit/Preferences/AudioPreferences.swift:37]: This code doesn't seem to belong to the Audio preference as it converts from ID to Reciter and `updateRecentRecitersList` seems to do the other way around. I suggest you either:
1) Move them to ReciterDataRetriever and probably rename that struct, or
2) Create RecentRecitersService that encapsulates this logic.

**Review by mohamede1945 — COMMENTED** [review]
Great work Zubair, just a small organizational change.

**Review by mohamede1945 — APPROVED** [review]
Great work, Jazak Allah khyran Zubair!
